File: HackerRank/Dictinary/LRU/LRU.cs
using System;
using System.Collections.Generic;

namespace LRU
{
    public class LRU<T>
    {
        private Dictionary<int, LinkedListNode<Tuple<T, int>>> _dictinaryQueckMemory;
        private LinkedList<Tuple<T, int>> _myQueckMemory;
        private int counter;
        public LRU(int length)
        {
            counter = length;
            _dictinaryQueckMemory = new Dictionary<int, LinkedListNode<Tuple<T, int>>>(counter);
            _myQueckMemory = new LinkedList<Tuple<T, int>>();
        }

        public T Get(int id)
        {
            if (_dictinaryQueckMemory.ContainsKey(id))
            {
                _myQueckMemory.AddFirst(_dictinaryQueckMemory[id].Value);
                _myQueckMemory.Remove(_dictinaryQueckMemory[id]);
                _dictinaryQueckMemory[id] = _myQueckMemory.First;
                return _dictinaryQueckMemory[id].Value.Item1;
            }
            else
            {
                return default(T);
            }
        }

        public void Push(int id, T value)   //dobavit' dannye po indexu
        {
            if (!_dictinaryQueckMemory.ContainsKey(id))
            {
                var newTuple = new Tuple<T, int>(value, id);
                _myQueckMemory.AddFirst(newTuple);
                LinkedListNode<Tuple<T, int>> node = _myQueckMemory.First;
                _dictinaryQueckMemory.Add(id, node);
            }
            else
            {
                LinkedListNode<Tuple<T, int>> node = _dictinaryQueckMemory[id];
                _myQueckMemory.AddFirst(node.Value); 
                _myQueckMemory.Remove(node);
                _dictinaryQueckMemory[id] = _myQueckMemory.First; //ukazyvaet na pervyj
            }

            if (_dictinaryQueckMemory.Count >= counter)  // get from constructor
            {
                int k = _myQueckMemory.Last.Value.Item2;
                _myQueckMemory.Remove(_myQueckMemory.Last);
                _dictinaryQueckMemory.Remove(k);
            }
        }
    }
}

File: HackerRank/Dictinary.Java/LRU/LRU.java
package LRU;

import java.util.*;

public class LRU<T>
{
	private HashMap<Integer, LinkedListNode<Tuple<T, Integer>>> _dictinaryQueckMemory;
	private LinkedList<Tuple<T, Integer>> _myQueckMemory;
	private int counter;
	public LRU(int length)
	{
		counter = length;
		_dictinaryQueckMemory = new HashMap<Integer, LinkedListNode<Tuple<T, Integer>>>();
		_myQueckMemory = new LinkedList<Tuple<T, Integer>>();
	}

	public final T Get(int id)
	{
		if (_dictinaryQueckMemory.containsKey(id))
		{
			_myQueckMemory.addFirst(_dictinaryQueckMemory.get(id).Value);
			_myQueckMemory.remove(_dictinaryQueckMemory.get(id));
			_dictinaryQueckMemory.put(id, _myQueckMemory.First);
			return _dictinaryQueckMemory.get(id).Value.Item1;
		}
		else
		{
			return null;
		}
	}

	public final void Push(int id, T value) //dobavit' dannye po indexu
	{
		if (!_dictinaryQueckMemory.containsKey(id))
		{
			Tuple<T, Integer> newTuple = new Tuple<T, Integer>(value, id);
			_myQueckMemory.addFirst(newTuple);
			LinkedListNode<Tuple<T, Integer>> node = _myQueckMemory.First;
			_dictinaryQueckMemory.put(id, node);
		}
		else
		{
			LinkedListNode<Tuple<T, Integer>> node = _dictinaryQueckMemory.get(id);
			_myQueckMemory.addFirst(node.Value);
			_myQueckMemory.remove(node);
			_dictinaryQueckMemory.put(id, _myQueckMemory.First); //ukazyvaet na pervyj
		}

		if (_dictinaryQueckMemory.size() >= counter) // get from constructor
		{
			int k = _myQueckMemory.getLast().Item2;
			_myQueckMemory.remove(_myQueckMemory.Last);
			_dictinaryQueckMemory.remove(k);
		}
	}
}